samples/hello-cube.html

```diff
@@ -0,0 +1,313 @@
+    /* Context, Device, SwapChain */
+    const adapter = await gpu.requestAdapter();
+    device = adapter.createDevice();
+
+    const canvas = document.querySelector('canvas');
+
+    const aspect = Math.abs(canvas.width / canvas.height);
+    mat4.perspective(projectionMatrix, (2 * Math.PI) / 5, aspect, 1, 100.0);
+
+    const context = canvas.getContext('gpu');
+
+    const swapChainDescriptor = {
+        context: context,
+        format: "bgra8unorm"
+    };
+
+    const swapChain = device.createSwapChain(swapChainDescriptor);
+
+    /* Shader Module */
+    const shaderModuleDescriptor = { code: shader };
+    const shaderModule = device.createShaderModule(shaderModuleDescriptor);
+
+    const verticesBufferDescriptor = {
+        size: vertexArray.byteLength,
+        usage: GPUBufferUsage.VERTEX | GPUBufferUsage.TRANSFER_DST
+    };
+    verticesBuffer = device.createBuffer(verticesBufferDescriptor);
+    verticesBuffer.setSubData(0, vertexArray.buffer);
+
+    /* Input State */
+    const positionAttributeDescriptor = {
+        attributeIndex: 0,
+        offset: 0,
+        format: "float4"
+    };
+    const colorAttributeDescriptor = {
+        attributeIndex: 1,
+        offset: colorOffset,
+        format: "float4"
+    }
+    const vertexBufferDescriptor = {
+        stride: vertexSize,
+        stepMode: "vertex",
+        attributes: [positionAttributeDescriptor, colorAttributeDescriptor]
+    };
+    const inputStateDescriptor = {
+        indexFormat: "uint32",
+        vertexBuffers: { 0: vertexBufferDescriptor }
+    };
+
+    /* Bind Group Layout */
+    const transformBufferBindGroupLayoutBinding = {
+        binding: 0,
+        visibility: GPUShaderStageBit.VERTEX,
+        type: "uniform-buffer"
+    };
+
+    const uniformsBindGroupLayoutDescriptor = {
+        bindings: [transformBufferBindGroupLayoutBinding]
+    };
+    uniformsBindGroupLayout = device.createBindGroupLayout(uniformsBindGroupLayoutDescriptor);
+
+    const depthStateDescriptor = {
+        depthWriteEnabled: true,
+        depthCompare: "less"
+    };
+
+    /* Render Pipeline */
+    const vertexStageDescriptor = {
+        module: shaderModule,
+        entryPoint: "vertex_main"
+    };
+    const fragmentStageDescriptor = {
+        module: shaderModule,
+        entryPoint: "fragment_main"
+    };
+    const pipelineDescriptor = {
+        bindGroupLayouts: [uniformsBindGroupLayout],
+
+        vertexStage: vertexStageDescriptor,
+        fragmentStage: fragmentStageDescriptor,
+
+        primitiveTopology: "triangle-list",
+        depthStencilState: depthStateDescriptor,
+        inputState: inputStateDescriptor
+    };
+    pipeline = device.createRenderPipeline(pipelineDescriptor);
+
+    /* Renderpass Attachments */
+    let colorAttachment = {
+        // attachment is acquired in render loop.
+        loadOp: "clear",
+        storeOp: "store",
+        clearColor: { r: 0.5, g: 1.0, b: 1.0, a: 1.0 }
+    };
+
+    // GPUExtent3D
```

Comment:
I wish we could do something more type-safe here. It seems silly to attach a 3D size to a 2D texture.

Reply:
Yeah, maybe we need GPUSize2D and GPUSize3D (I prefer both of them over "Extent")